### ISSUE TITLE
runtime: leaking nested goroutines

### ISSUE THREAD

**Issue by nuivall** [golang/go#16022]
Hello,

1. What version of Go are you using (`go version`)?
go version devel +09eedc3 Wed Jun 8 05:24:53 2016 +0000 darwin/amd64
tested also on 1.5 and 1.6 with the same results

3. What did you do?
https://play.golang.org/p/ATENOgjVp-

4. What did you expect to see?
Program waits forever but is holding small amount of memory.

5. What did you see instead?
Program holds ~1.3GB of memory, after forcing GC in Point B nothing changes while GC in Point A makes program using only 22MB.

go tool pprof http://localhost:6060/debug/pprof/heap reports less memory usage but still indicates a leak, it is showing that nearly all memory is taken by descendants of runtime.newproc1.

I'm looking forward to answer any questions or hear why observed behaviour is expected :)

**Comment by ianlancetaylor** [comment]
When you say that the program holds ~1.3GB of memory, how are you measuring that?

Have you tried calling https://golang.org/pkg/runtime/debug/#FreeOSMemory ?

**Comment by randall77** [comment]
I can reproduce this on Linux with GOMAXPROCS=1 on go1.6.

What is happening is that you are spawning 600,000 goroutines. Depending on how the scheduler behaves, if all are spawned before any finish, you'll use up to ~1.3GB (~2K per goroutine).  If goroutines finish faster than you spawn them, then memory usage will be quite low.  That's why GOMAXPROCS matters, and why the GC at point A affects things.

When you get to the sleep, Go is not using the ~1.3GB anymore.  But it hasn't given that memory back to the OS yet.  Wait 5 minutes or call debug.FreeOSMemory.

**Comment by adg** [comment]
I believe this is the expected behavior and that this issue should be closed.

**Comment by nuivall** [comment]
Thank you for your responses! I tried FreeOSMemory (btw small nit in the code I posted, I should wrap fmt.Println with a function), I waited 20 minutes, previously I also made system swapping. Nothing changed. I understand that Go likes to stash memory for later but how this pprof output can be explained, why it reports 168MB in runtime.malg 384B bucket (/debug/pprof/goroutine shows only 6 goroutines)?

![pprof_tf 3](https://cloud.githubusercontent.com/assets/2939488/15958139/13627906-2ef4-11e6-901d-12ba22144abe.png)
